### ISSUE TITLE
Plotting Library?

### ISSUE THREAD

**Issue by aria42** [clojupyter/clojupyter#51]
Do we have plans around adding a plotting library? I was thinking we could incorporate one and extend the MIME protocol to it to directly render either a PNG or SVG or whatever yields the sanest result. 

I'm up for doing this if it seems valuable. Is it crazy to just include the plotting as a dependency so people can require in the kernel? If not, we need to first solve the issue around getting a classpath to the kernel. I saw in another issue from @roryk that he wanted to re-use the cider-middleware. I'm not sure if that would work because most of cider assumes you're launching from the `project.clj` and you just inherit it's classpath (once that's setup you can send `:load-file` op to nrepl with the middleware.

**Comment by luojieshun** [comment]
Plotting already works with incanter. You can can new dependencies in the
notebook as well.

**Comment by aria42** [comment]
Do you have a sample notebook somewhere? My dependencies in local maven repo don't seem to load.

**Comment by luojieshun** [comment]
I pushed a notebook in my fork

https://github.com/luojieshun/clojupyter/blob/example/examples/Incanter%20Plot.ipynb

**Comment by aria42** [comment]
That's awesome, I didn't notice that helper, we should add to `Readme.md` and also your sample. It'd be great if you can PR your example, I'm happy to as well.

**Comment by didiercrunch** [comment]
I have tried many plotting libraries.

I wanted to use plotly but their apis are not very well documented. In a
very strange manner, there is a very nice api spec but the endpoint which
consumes it is hidden.

I have used  highchart, a client side library,  and it worked pretty well
but github is not showing the charts.  This is a problem shared by any
library that renders chart as html (javascript, css...,) and not plain old
svg or png.

Geom is a very interresting library which I had great success.  Geom api is
a bit complicates but powerfull.  The source code itself is strange as
everything is built with orgmode.

I have never tried incanter but I will give it a try.

In conclusion, I believe we need to document how to plot in clojupyter.

Cheers


Here is a repo were I made my experiments: https://github.
com/didiercrunch/graphl

**Comment by pdenno** [comment]
I also think it would be helpful to have @luojieshun 's example (screenshot) in the README. It demonstrates the progress being made in Clojupyer. Collecting examples in a directory is a great idea.

**Comment by luojieshun** [comment]
Incanter is more like a R/Pandas replacement rather then a plotting
library. I believe it uses jfreechart internally. IPython has tight
integration with matplotlib, I have not found anything comparable to it in
Clojure/Java yet.

**Comment by aria42** [comment]
My bigger concern is that incanter is kinda unsupported at this point and the API is a little crufty (but still functional). This might be a little too ambitious, but I've seen seen lots of examples of people using d3 in Jupyter and generating HTML Python-side for d3.

**Comment by didiercrunch** [comment]
Interestingly, matplotlib is not a vip library for jupyter; it uses the
same feature as we are.  The python kernel adds special syntax (the % and
%%) which monkey patch the python's std.

In our case, monkey patching is not an option.  But we could make a macro
like 'defchart' which would wrap a charting library and the clojupyter
functionnalities.

It is impressive what geom can do.  Just take a look at
https://github.com/didiercrunch/graphly/blob/master/geom.ipynb (not from a
cell phone) and you will see results very similar to what matplotlib does.

As a proof of concept, we could implement that same charting capabilities
than gorilla backed by geom and see if it can work.

**Comment by luojieshun** [comment]
Wow those geom plot looks great!

**Comment by aria42** [comment]
Monkey patching isn't an option, but we can dynamically extend a protocol to a class if it returns a particular class (record, type, etc.). What's there in `helper` namespace is good too but we can make it a little easier maybe by just having something apply hiccup to generate html so there's fewer conversion steps.

**Comment by aria42** [comment]
Actually what do people think about this design principle: If something affects the rendering of the cell (adding JS file or displaying html) we should use cell magic, rather than have people require a name-space and call a funciton. 

I can see arguments either way, but thought I'd get people's thoughts.

**Comment by metasoarous** [comment]
For the record, [Oz](https://github.com/metasoarous/oz) features a `oz.notebook.clojupyter` namespace which lets you use Vega & Vega-Lite from Clojupyter.
